### PR TITLE
Add support for TFTP related DHCP option

### PR DESCRIPTION
**Pull request by HayleyDeckers**
Adds support for 3 TFTP related options (`TFTServerName`, `TFTPServerAdress`, and `BootfileName`) which can be used to boot devices from the network. 

The capitalization of `BootfileName` follows the RFC and IANA spelling, it is different from the existing `BootFileSize` but that is consistent with the spec 😅

### CODE REVIEW

File: src/v4/options.rs [hunk 1]
```diff
@@ -1199,7 +1217,8 @@ impl Encodable for DhcpOption {
             | RouterSolicitationAddr(addr)
             | RequestedIpAddress(addr)
             | ServerIdentifier(addr)
-            | SubnetSelection(addr) => {
+            | SubnetSelection(addr)
+            | TFTPServerAdress(addr) => {
```
Comment (stappersg) [inline src/v4/options.rs:1221]: That
```text
-         | SubnetSelection(addr) => {
+         | SubnetSelection(addr)
+         | TFTPServerAdress(addr) => {
```
how that curly bracket open, `{`, did **not** stay with the `SubnetSelection()` feels strange.

**Comment by leshow** [comment]
This might be pedantic but, do we _know_ the server name and bootfile name are able to be utf-8 encoded? We originally had `sname` and `fname` as `String` but ended up changing them after figuring out the protocol only really ensure they are a bunch of bytes.

**Comment by HayleyDeckers** [comment]
> We originally had sname and fname as String but ended up changing them after figuring out the protocol only really ensured they are a bunch of bytes.

According to https://www.rfc-editor.org/rfc/rfc2131 and https://www.rfc-editor.org/rfc/rfc951 `sname` and `file` are null-terminated strings. The encoding is not given, but (NVT) ASCII seems logical to assume given the age and other references to it.

I think it's safe to assume those fields and these options should be valid utf-8 (though not every utf-8 string is a valid option), the octets are also indicated as `c1,c2,c3,..,cn` suggesting that they're meant to be interpreted as characters _but_ I'll agree the spec is not very clear.

**Comment by leshow** [comment]
For reference, here is the issue that caused the change to begin with: https://github.com/bluecatengineering/dhcproto/issues/3 do you have any additional thoughts?

**Comment by HayleyDeckers** [comment]
I think `bstr` might be the right call as 

> [this crate enables] their use as byte strings, where byte strings are conventionally UTF-8. 

likely matches reality, whatever the standard-as-written says. Writing utf-8 will always be valid (utf-8 can't have interior null-bytes), but the full encoding seems undefined aside from fname/sname having to be a "null terminated string".

**Comment by leshow** [comment]
OK, as long as no traits need to be put in the public API, I think `BString` makes sense

File: src/v4/options.rs (hunk 1 above)
Comment (leshow) [reply src/v4/options.rs:1251]: As long as it's gone through rustfmt!

**Comment by HayleyDeckers** [comment]
Hi, happy new year :)

I had another look at this and I think the issue of unclear encodings goes beyond just these couple of TFTP-related options. Take for example Host name option 12, for which the [rfc](https://www.rfc-editor.org/rfc/rfc2132.html) says
> See [RFC 1035](https://www.rfc-editor.org/rfc/rfc1035) for character set restrictions. 

Which in turn mentions a _preferred_ format
> Although labels can contain any 8 bit values in octets that make up a label, it is strongly recommended that labels follow the preferred syntax described elsewhere in this memo

This might need to be a BString too as String could be to restrictive. 

On the other hand `MeritDumpFile`, `RootPath`,  `NISDomain`, `Message`, take a `String` to but this is too permissive as the RFC says these use NVT ASCII.

`NetBiosScope` I'm not quite sure what the format should be. 

---

I don't might changing the Strings to BString (I did so just now), but it might make sense to merge this PR as is and punt the resolution of the types to another PR or Issue to make sure they are kept consistent?

**Comment by leshow** [comment]
Hey! If we're to be consistent, from my reading of the RFC opt 66/67 contain the data from `sname` and `fname`.

>    This option is used to identify a TFTP server when the 'sname' field
   in the DHCP header has been used for DHCP options.

Right now those fields are `Vec<u8>`, so maybe we should just stick with that for now? I think `BString` is a fine choice also although I do prefer std types where possible.

In general, I think being more permissive than is strictly necessary is okay, I'd rather err of the side of the parser not blowing up. When we know the contents are ASCII (as in `Message`), I was thinking we would just use `String`, unless there is something to be gained by using a different type there? 

Which reminds me, we also don't handle ['option overload'](https://www.rfc-editor.org/rfc/rfc2132.html#section-9.3).  If set, the `sname` and `fname` could contain option data, the part that's usually added in the variable length section at the end of the packet. I don't think option overload is used much these days, but it is another condition that we don't cover.

Sorry, lot of words to discuss encodings! Thanks for following up on this.

**Comment by HayleyDeckers** [comment]
> Hey! If we're to be consistent, from my reading of the RFC opt 66/67 contain the data from sname and fname.

That makes sense looking at the quote you gave though as a side-note, the maximum length of field 66/67 is 256 characters and must not be null-terminated, whereas `sname` [must be a null-terminated string of at most 64 characters (including null)](https://www.rfc-editor.org/rfc/rfc2131).

> This option is used to identify a TFTP server when the 'sname' field
in the DHCP header has been used for DHCP options.

I'm not quite sure how to interpret this line from the RFC, but I know there are embedded devices (raspberry pis for example) that will accept passing the tftp-server in Option 66, but will not accept passing it in `sname`. So in practice I think it makes sense to consider them as two different settings. 

> In general, I think being more permissive than is strictly necessary is okay, I'd rather err of the side of the parser not blowing up. When we know the contents are ASCII (as in Message), I was thinking we would just use String, unless there is something to be gained by using a different type there?

For accepting messages from other devices, yes definitely. Not all devices will follow the letter of the RFC unfortunately. 
For sending I think it's less straightforward to say what is or isn't acceptable. If the RFC says that an option should be ASCII, and you accept a String in the constructor, you could create an invalid value by passing `"🦀".to_string()`. But then, maybe you want to be able to do that? 

----

I'll convert this field to be a `Vec<u8>` instead of a String for maximum flexibility for now.

**Comment by leshow** [comment]
> For accepting messages from other devices, yes definitely. Not all devices will follow the letter of the RFC unfortunately.
For sending I think it's less straightforward to say what is or isn't acceptable. If the RFC says that an option should be ASCII, and you accept a String in the constructor, you could create an invalid value by passing "🦀".to_string(). But then, maybe you want to be able to do that?

Good point, I was thinking of decoding not encoding. I will make an issue to address the ASCII types. Thank you for following up again, if you switch to `Vec<u8>` for the new types in this PR I will merge.

**Comment by HayleyDeckers** [comment]
>  if you switch to Vec<u8> for the new types in this PR I will merge.

done! :)

> My interpretation was that it relates to when option overload is used, in this case the sname field contains DhcpOptions and the TFTP server option will house the value that was in sname?

I read it as "`sname` will hold the server-name of the `tftp-server`, unless option Overload is used, in which case Option 66 can be used to store the server-named".  

In the old BOOTP protocol, `sname` and `fname` full-filled similar purposes to TFTP so I think that makes sense.  See also [RFC  5859](https://www.rfc-editor.org/rfc/rfc5859) which says

>  There are two commonly accepted methods to discover this
   server via DHCP; the "sname" field in the DHCP header [[RFC2131](] and
   the "TFTP Server Name" option (66) [[RFC2132](https://www.rfc-editor.org/rfc/rfc2132)].  Both of these sources of information, however, contain the TFTP server's hostname.  That
   hostname must then be translated to an IP address.  The usual method
   to accomplish this would be DNS [[RFC1034](https://www.rfc-editor.org/rfc/rfc1034)].
   
TFTP is often used to bootstrap embedded devices, with the firmware responsible for handling the protocol and, unfortunately,  that likely means there are a lot of quirks and deviations from the standard in the real-world.

I've noticed however that some devices will only accept `option 66` and will not look at `sname`.

**Pull request merged**